#Analyzing the NYC Subway Dataset

##Background Information

In this project, we look at the NYC Subway data and figure out if more people ride the subway when it is raining versus when it is not raining.
The dataset that is available contains data about turnstiles at different stations as well as weather conditions.

##Statistical Test

###Independent, dependent variables

First, let's see what kind of data is in the dataset by exploring the column names.

In [2]:
import pandas as pd
filename='data/improved-dataset/turnstile_weather_v2.csv'
turnstile_weather = pd.read_csv(filename)
print list(turnstile_weather.columns.values)

['UNIT', 'DATEn', 'TIMEn', 'ENTRIESn', 'EXITSn', 'ENTRIESn_hourly', 'EXITSn_hourly', 'datetime', 'hour', 'day_week', 'weekday', 'station', 'latitude', 'longitude', 'conds', 'fog', 'precipi', 'pressurei', 'rain', 'tempi', 'wspdi', 'meanprecipi', 'meanpressurei', 'meantempi', 'meanwspdi', 'weather_lat', 'weather_lon']


The purpose of this project is to find out if more people use the subway when it is raining. Fortunately, there is a column rain that indicates if it was raining at a certain time and location. According to the documentation it is an indicator (0 or 1). 
How to measure if the subway is used more? Well, the data contains the number of entries at some units where a unit is a data collector of multiple turnstiles. ENTRIESn is the reading of cummulative turnstile entries from the remote unit and ENTRIESn_hourly indicates the difference in ENTRIES from the previous REGULAR reading. It seems like ENTRIESn_hourly is a good measure of how many people will use the subway.

Therefore, the independent variable is the rain. It can either be a 0 (no rain) or 1 (rain)
The dependent variable is the the number of entries since the last reading (ENTRIESn_hourly)

### Hypothesis and chosen test

The data that is provided is a sample. With the hypothesis tests we want to explore if we can conclude that the subway is used more when it is raining. 
Let μR and μN respectively be the population means for ENTRIESn_hourly when it is not raining and when it is raining. Let H0 and Ha respectively be the 0 and alternative hypothesis. The hypotheses can be written as:

H0 : μR = μN

Ha : μR > μN

When comparing means of populations, multiple tests can be used. A t-test or z-test can be used if the data is assumed to be normally distributed. 
However, the data from our sample is not normally distributed if we look at the plot below. Mann–Whitney´s U test is a nonparametrict test that can be used if we don't assume that the data follows a certain distribution. Because we don´ t assume any underlying distribution this test will be used. The test will be one tailed, because it is only interesting to know if **more** people use the subway when it is raining. The p critical value is 0.05, since this is quite common.

![Histogram of rain vs norain values](plots/histogram_rain_norain.png)

### Results from executing the test

Below the results from the test can be found.

In [13]:
import numpy as np
import scipy.stats
rain = turnstile_weather['ENTRIESn_hourly'][turnstile_weather['rain'] == 1]
norain = turnstile_weather['ENTRIESn_hourly'][turnstile_weather['rain'] == 0]
print "length norain df vs rain df:", len(norain), len(rain)
print "Norain mean + std:", np.mean(norain), np.std(norain)
print "Rain mean + std:", np.mean(rain), np.std(rain)
U, p = scipy.stats.mannwhitneyu(rain, norain)
print "U-statistic:", U
print "p-value:", p

length norain df vs rain df: 33064 9585
Norain mean + std: 1845.53943866 2878.72731433
Rain mean + std: 2028.19603547 3189.26699226
U-statistic: 153635120.5
p-value: 2.74106957124e-06


The mannwhitney u test is per default one-tailed.  As can be seen, the p-value is a lot smaller than the p critical value of 0.05. Therefore it is very unlikely that the population means are equal, so the alternative hypothesis applies: more people ride the subway when it is raining.

##Linear Regression

Machine learning is a branch of artificial intelligence focused on constructing systems that learn from large amounts of data to make predictions.
Linear Regression is one of the most used models in machine learning.

### Regression model
To create a lineair regression model I will use the Ordinary Least Squares (OLS) from Statsmodels. 

### Features 

I have used the following features in the regression model

* rain: because in the statistical test, it is clear that rain matters.
* hour: because there are probably more people taking the subway in peak hours.
* meantempi: because there are probably less travelers when it is very cold or hot.
* weekday: because I think some days are more populair to work than others.
* UNIT (dummy variable): because when I used this, it drastically improved my R2 value.

### Parameters of non-dummy features and R-squared

below the parameters of the non-dummy features and R-squared are shown

In [11]:
import matplotlib.pyplot as plt
import statsmodels.api as sm

def linear_regression(features, values):
    """
    Perform linear regression given a data set with an arbitrary number of features. 
    This can be the same code as in the lesson #3 exercise.
    """ 
    features = sm.add_constant(features)
    model = sm.OLS(values, features)
    results = model.fit()
    intercept = results.params[0]
    params = results.params[1:]    
    return intercept, params
    

def compute_r_squared(data, predictions):
    '''
    Calculates R sqared based on given data and predictions
    '''
    SST = ((data - np.mean(data)) **2).sum()
    SSREG = ((predictions - data) **2).sum()
    r_squared = 1 - SSREG / SST
    
    return r_squared

filename='data/improved-dataset/turnstile_weather_v2.csv'
turnstile_weather = pd.read_csv(filename)
features = turnstile_weather[['rain', 'hour', 'meantempi', 'weekday']]
dummy_units = pd.get_dummies(turnstile_weather['UNIT'], prefix='unit')
features = features.join(dummy_units)
values = turnstile_weather['ENTRIESn_hourly']
intercept, params = linear_regression(features, values)
predictions = intercept + np.dot(features, params)
print "param-rain", params['rain']
print "param-hour", params['hour']
print "param-meantempi", params['meantempi']
print "param-weekday", params['weekday']
print "r-squared:", compute_r_squared(values, predictions)

param-rain -38.034135646
param-hour 123.403811602
param-meantempi -13.8580644847
param-weekday 982.860016244
r-squared: 0.482340215273


The R2 value is about 0.48, so what does this mean? R2 is a measure of how well the data fits the regression model. The value of R2 is 1 for a perfect fit and 0 when there is no fit, so in this case the fit is reasonable.

##Visualization

Below a visualization can be found about the ridership by time of day. On the x-axis the hour is give and on the y-axis the number of entries.

![Histogram of rain vs norain values](plots/number_of_entries_per_hour.png)


## Conclusion

#### Do more people ride the NYC subway when it is raining or when it is not raining?

From the results obtained from the statistical test and lineair regression, we can say in general that more people ride the NYC subway when it is raining. However, rain is not a very good predictor.

In the statistical test, the conclusion is that our sample supports the alternative hypothesis, the population mean for rain is greater that for non rain.

In our lineair regression model, the influence of the rain on the ouput is negative. I have also tried to leave out the feature rain and it almost didn't change the R2 value. The purpose of the regression model is to be able to predict the outcome based on certain features. However, these features influence each other. If one is left out, it will change the param values of the other features. The dummy variable UNIT has a big influence on R2. This leads to the conclusion that UNIT is a good predictor, but rain isn't.


## Reflection

#### Shortcomings

There are some shortcomings in my analysis.
* dataset: the dataset contains a lot more rows for non-rain than for rain. It would be ideal to have equal number of rows for both cases.
* regression model: the chosen regression model OLS (ordinary least squares) is quite simplistic and has some drawbacks. It is sensitive for outliers and as seen in the figure of entries per hour, there are some outliers. Next to this, OLS performs bad when features are put into the model, that have a dependence between them. 

